DECISION-RESEARCH.md

```diff
@@ -0,0 +1,315 @@
+- Explicit approach with predefined available actions
+- Allows appending to arrays
+- Multiple operations are performed by sending an array of “operation” objects
+
+#### RFC 7396 (merge-patch+json)
+
+- Straightforward, minimal verbosity
+- Removal is “implicit”
+  - i.e. by sending a key with a `null` value
+- No support for extending array values 
+  - i.e. arrays must be fully replaced
+- PHP tooling seems to be scarce and mostly abandoned
+- Multiple operations are performed by sending one “update” object
+
+---
+
+## Entities + Statements
+
+**Decision: Statements would be exposed both within an entity and also at the top level**
+
+Considered options:
+
+- For the statements with a GUID:
+  - Should we address those always through entities?
+    - longer path
+    - the entity id is already in the path and the prefix is redundant
+- Or individually, through a /statements path? 
+
+Usefulness (end users) Vs. cost of ownership (devs)
+
+### Notes
+
+The current Action API allows you to get a statement with only the GUID (without the entity id prefix) with no extra work
+
+Not having a top level statement endpoint would mean that “extra work” would be needed by the client in order to access a statement (splitting the statement ID)
+
+Taking this into account we can cover all of our existing APIs with REST APIs only if we do have statements as a top level api module.
+
+---
+
+## RevisionIds (Path? Query?)
+
+**Decision: We will not worry about revisionids in paths, as the current API doesn't expose them and per core behaviour this should probably be left for the revisions endpoint**
+
+Both choices seem to be okay according to stackoverflow https://stackoverflow.com/a/12737540/4746236
+
+Mediawiki core REST API only has revisionid when looking specifically at revisions (not pages) https://www.mediawiki.org/wiki/API:REST_API
+
+- In the query:
+  - Drupal https://www.drupal.org/docs/core-modules-and-themes/core-modules/jsonapi-module/revisions
+- In the Path:
+  - Wikimedia (RestBase) https://en.wikipedia.org/api/rest_v1/#/Page%20content/getFormatRevision
+  - Wordpress https://developer.wordpress.org/rest-api/reference/page-revisions/#retrieve-a-page-revision
+
+---
+
+## Conditional Requests
+
+**Decision: We will implement conditional requests for gets.**
+
+- TL;DR use HTTP spec to prevent the “lost update problem”
+- MediaWiki core REST API makes use of conditional requests
+- ETag + HTTP 409 CONFLICT
+- Preconditions in headers
+  - If-Match
+  - If-Unmodified-Since
+- Should happen for all sub-paths of entities (which all rely on entity state)
+- (Possibly even require it via HTTP 428 PRECONDITION REQUIRED)
+
+Sources:
+
+- https://dzone.com/articles/concurrency-control-in-rest-api-with-spring-framew
+- https://www.novatec-gmbh.de/en/blog/managing-concurrency-in-a-distributed-restful-environment-with-spring-boot-and-angular2/
+
+---
+
+## Resources in Plural vs Singular
+
+**Decision: Plural for everything.**
```

Comment:
```suggestion
**Decision: Plural for everything. For consistency's sake.**
```

or mention somewhere that this (not some preference) was the compelling reason for us to make the call.